src/Inline/Element/AbstractInline.php

```diff
@@ -26,38 +26,11 @@
     public $data = [];
 
     /**
-     * @var AbstractInline|null
+     * @return bool
      */
-    protected $parent;
-
-    /**
-     * @return AbstractBlock|AbstractInline|null
-     */
-    public function getParent()
+    public function isContainer()
     {
-        return $this->parent;
-    }
-
-    /**
-     * @param AbstractBlock|AbstractInline $parent
-     *
-     * @return $this
-     */
-    public function setParent($parent)
-    {
-        if (!$parent instanceof AbstractBlock && !$parent instanceof self) {
-            throw new \InvalidArgumentException(sprintf(
-                'Argument 1 passed to %s() must be an instance of %s or %s, instance of %s given',
-                __METHOD__,
-                'League\CommonMark\Block\Element\AbstractBlock',
-                'League\CommonMark\Block\Element\AbstractInline',
-                is_object($parent) ? get_class($parent) : gettype($parent)
-            ));
-        }
-
-        $this->parent = $parent;
-
-        return $this;
+        return false;
```

Comment:
> there is no way implement it without multiple inheritance... May be you have some thoughts about this problem?

Yeah, I encountered this issue myself a long time ago when I first tried making these changes.

Maybe we could replace `AbstractInline` and `AbstractBlock` with interfaces?  For example:

~~~php
class Node {
    // ...
}

class ContainerNode extends Node {
    public function appendChild() { }
    public function prependChild() { }
    public function getChildren() { }
    public function getFirstChild() { }
    public function getLastChild() { }
    public function detachChildren() { }
    public function replaceChildren() { }
}

interface BlockElementInterface {
    // All of the abstract functions currently found in AbstractBlock:
    public function canContain(Node $node);
    public function acceptsLines();
    public function isCode();
    public function matchesNextLine();
    public function handleRemainingContents();
}

interface InlineElementInterface {
    // TODO: Is this interface even necessary?
}

// This replaces AbstractStringContainer
interface StringContainerInterface extends InlineElementContainer {
    public function getContent();
    public function setContent();
}

// This replaces AbstractWebResource
interface WebResourceInterface extends InlineElementContainer {
    public function getUrl();
}

~~~

(I have removed the method parameters above to keep the example simple)

You'd therefore end up combining them like so:

~~~php
class Paragraph extends ContainerNode implements BlockElementInterface { }

class HorizontalRule extends Node implements BlockElementInterface { }

class Emphasis extends ContainerNode implements InlineElementInterface { }

class Newline extends Node implements InlineElementInterface { }

class Text extends Node implements StringContainerInterface { }
~~~

If you need to know whether something is a block or inline, check for the relevant interface.

Do you think something like that would work?